src/tree.rs

```diff
@@ -126,12 +128,156 @@ pub fn minimum_spanning_tree(
+/// the population target within some epsilon. The function returns a list of
+/// all such possible cuts, represented as the set of nodes in one
+/// partition/subtree. Wraps around ``_bipartition_tree``.
+///
+/// :param PyGraph graph: Spanning tree. Must be fully connected
+/// :param pops: The populations assigned to each node in the graph.
+/// :param float pop_target: The population target to reach when partitioning the
+///     graph.
+/// :param float epsilon: The maximum percent deviation from the pop_target
+///     allowed while still being a valid balanced cut edge.
+///
+/// :returns: A list of tuples, with each tuple representing a distinct
+///     balanced edge that can be cut. The tuple contains the root of one of the
+///     two partitioned subtrees and the set of nodes making up that subtree.
+#[pyfunction]
+#[pyo3(text_signature = "(spanning_tree, pop, target_pop, epsilon)")]
+pub fn bipartition_tree(
+    _py: Python,
+    spanning_tree: &graph::PyGraph,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> PyResult<Vec<(usize, Vec<usize>)>> {
+    Ok(_bipartition_tree(spanning_tree, pops, pop_target, epsilon))
+}
+
+/// Internal _bipartition_tree implementation.
+fn _bipartition_tree(
+    spanning_tree: &graph::PyGraph,
+    pops: Vec<f64>,
+    pop_target: f64,
+    epsilon: f64,
+) -> Vec<(usize, Vec<usize>)> {
+    let mut pops = pops;
+    let spanning_tree_graph = &spanning_tree.graph;
+    let mut same_partition_tracker: Vec<Vec<usize>> =
+        vec![vec![]; spanning_tree_graph.node_bound()]; // Keeps track of all all the nodes on the same side of the partition
+
+    let mut node_queue: VecDeque<NodeIndex> = VecDeque::<NodeIndex>::new();
+    for leaf_node in spanning_tree_graph.node_indices() {
+        if spanning_tree_graph.neighbors(leaf_node).count() == 1 {
+            node_queue.push_back(leaf_node);
+        }
+        same_partition_tracker[leaf_node.index()].push(leaf_node.index());
+    }
+
+    // BFS search for balanced nodes
+    let mut balanced_nodes: Vec<(usize, Vec<usize>)> = vec![];
+    let mut seen_nodes = HashSet::with_capacity(spanning_tree_graph.node_count());
+    while !node_queue.is_empty() {
+        let node = node_queue.pop_front().unwrap();
+        let pop = pops[node.index()];
+
+        // Mark as seen; push to queue if only one unseen neighbor
+        let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+            .graph
+            .neighbors(node)
+            .filter(|node| !seen_nodes.contains(&node.index()))
+            .collect();
+
+        if unseen_neighbors.len() == 1 {
+            // This will be false if root
+            let neighbor = unseen_neighbors[0];
+            pops[neighbor.index()] += pop;
+            let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
+            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+
+            if !node_queue.contains(&neighbor) {
```

Comment:
This will have bad performance since we need to traverse the queue to find `neighbor`. I'm wondering if we can invariably push `neighbor` to the queue and skip the logic inside the while loop if `node` is marked as seen.

Reply:
That's a good point! Implementing and benchmarking now.